flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/optimize/StreamNonDeterministicUpdatePlanVisitor.java

```diff
@@ -0,0 +1,1009 @@
+                nonDeterministicCols.put(index, ndCall.get());
+            } // else ignore
+        }
+        List<Integer> unsatisfiedColumns =
+                requireDeterminism.toList().stream()
+                        .filter(index -> nonDeterministicCols.containsKey(index))
+                        .collect(Collectors.toList());
+        if (!unsatisfiedColumns.isEmpty()) {
+            String errorMsg =
+                    generateNonDeterministicColumnsErrorMessage(
+                            unsatisfiedColumns,
+                            relatedRel.getRowType(),
+                            relatedRel,
+                            nonDeterministicCols,
+                            Optional.empty());
+            throw new TableException(errorMsg);
+        }
+    }
+
+    private void checkNonDeterministicCondition(
+            final RexNode condition, final StreamPhysicalRel relatedRel) {
+        Optional<String> ndCall = getNonDeterministicCallName(condition);
+        if (ndCall.isPresent()) {
+            throw new TableException(
+                    generateNonDeterministicConditionErrorMessage(
+                            ndCall.get(), condition, relatedRel));
+        }
+    }
+
+    private void checkUnsatisfiedDeterminism(
+            final ImmutableBitSet requireDeterminism,
+            final int aggStartIndex,
+            final List<AggregateCall> aggCalls,
+            final RelDataType rowType,
+            final StreamPhysicalRel relatedRel) {
+        Map<Integer, String> nonDeterministicOutput = new HashMap();
+        // skip checking non-deterministic columns in grouping keys or filter args in agg call
+        // because they were pushed down to input project which processes input only message
+        int aggOutputIndex = aggStartIndex;
+        for (AggregateCall aggCall : aggCalls) {
+            if (!aggCall.getAggregation().isDeterministic()
+                    || aggCall.getAggregation().isDynamicFunction()) {
+                nonDeterministicOutput.put(aggOutputIndex, aggCall.getAggregation().getName());
+            }
+            aggOutputIndex++;
+        }
+        // check if exist non-deterministic aggCalls which were in requireDeterminism
+        List<Integer> unsatisfiedColumns =
+                requireDeterminism.toList().stream()
+                        .filter(index -> nonDeterministicOutput.containsKey(index))
+                        .collect(Collectors.toList());
+        if (!unsatisfiedColumns.isEmpty()) {
+            String errorMsg =
+                    generateNonDeterministicColumnsErrorMessage(
+                            unsatisfiedColumns,
+                            rowType,
+                            relatedRel,
+                            nonDeterministicOutput,
+                            Optional.empty());
+            throw new TableException(errorMsg);
+        }
+    }
+
+    private String generateNonDeterministicConditionErrorMessage(
```

Comment:
throw TableException directly in this method and rename it throwNonDeterministicConditionError ?